src/main/scala/mesosphere/marathon/core/heartbeat/MesosHeartbeatMonitor.scala

```diff
@@ -0,0 +1,160 @@
+// @param heartbeatActor is the receipient of generated Heartbeat.Message's
+//
+// @see mesosphere.util.monitor.HeartbeatMonitor
+// @see org.apache.mesos.Scheduler
+// @see org.apache.mesos.SchedulerDriver
+class MesosHeartbeatMonitor @Inject() (
+    @Named(MesosHeartbeatMonitor.BASE) scheduler: Scheduler,
+    @Named(ModuleNames.MESOS_HEARTBEAT_ACTOR) heartbeatActor: ActorRef
+) extends Scheduler {
+
+  import MesosHeartbeatMonitor._
+
+  private[this] val log = LoggerFactory.getLogger(getClass.getName)
+
+  log.debug(s"created mesos heartbeat monitor for scheduler $scheduler")
+
+  protected[marathon] def heartbeatReactor(driver: SchedulerDriver): Heartbeat.Reactor = new Heartbeat.Reactor {
+    // virtualHeartbeatTasks is sent in a reconciliation message to mesos in order to force a
+    // predictable response: the master (if we're connected) will send back a TASK_LOST because
+    // the fake task ID and agent ID that we use will never actually exist in the cluster.
+    // this is part of a short-term workaround: will no longer be needed once marathon is ported
+    // to use the new mesos v1 http API.
+    private[this] val virtualHeartbeatTasks: java.util.Collection[TaskStatus] = Seq(fakeHeartbeatStatus).asJava
+
+    override def onSkip(): Unit = {
+      log.debug("Prompting mesos for a heartbeat via explicit task reconciliation")
+      driver.reconcileTasks(virtualHeartbeatTasks)
+    }
+
+    override def onFailure(): Unit = {
+      log.warn("Too many subsequent heartbeats missed; inferring disconnected from mesos master")
+      disconnected(driver)
+    }
+  }
+
+  override def registered(
+    driver: SchedulerDriver,
+    frameworkId: FrameworkID,
+    master: MasterInfo): Unit = {
+    log.debug("registered heartbeat monitor")
+    heartbeatActor ! Heartbeat.MessageActivate(heartbeatReactor(driver), sessionOf(driver))
+    scheduler.registered(driver, frameworkId, master)
+  }
+
+  override def reregistered(driver: SchedulerDriver, master: MasterInfo): Unit = {
+    log.debug("reregistered heartbeat monitor")
+    heartbeatActor ! Heartbeat.MessageActivate(heartbeatReactor(driver), sessionOf(driver))
+    scheduler.reregistered(driver, master)
+  }
+
+  override def resourceOffers(driver: SchedulerDriver, offers: java.util.List[Offer]): Unit = {
+    heartbeatActor ! Heartbeat.MessagePulse
+    scheduler.resourceOffers(driver, offers)
+  }
+
+  override def offerRescinded(driver: SchedulerDriver, offer: OfferID): Unit = {
+    heartbeatActor ! Heartbeat.MessagePulse
+    scheduler.offerRescinded(driver, offer)
+  }
+
+  override def statusUpdate(driver: SchedulerDriver, status: TaskStatus): Unit = {
+    heartbeatActor ! Heartbeat.MessagePulse
+
+    // filter TASK-LOST updates for our fake tasks to avoid cluttering the logs
+    if (!isFakeHeartbeatUpdate(status)) {
+      scheduler.statusUpdate(driver, status)
+    } else {
+      log.info("received fake heartbeat task-status update")
```

Comment:
imo debug would be enough. for long running services we expect only few updates (unless there are mesos health checks reporting stuff anyways)